### ISSUE TITLE
In CXXRTL, in some cases, always_ff blocks are evaluated despite the sensitivity conditions not being met

### ISSUE THREAD

**Issue by flaviens** [YosysHQ/yosys#4149]
### Version

Yosys 0.37+21 (git sha1 3d9e44d18, clang 10.0.0-4ubuntu1 -fPIC -Os)

### On which OS did this happen?

Linux

### Reproduction Steps

Hi there!

After #4146 I have been running designs with two calls to `step()`, and noticed a new issue.
It seems that in some cases, always_ff blocks are evaluated despite the sensitivity conditions not being met.

[Here is a repository to help you reproduce the issue](https://github.com/flaviens/cxxrtl-b5). The repository also contains a test for Verilator for comparison to convince myself that this is correct.

Please let me know if you need any help reproducing the issue, or if you need any more information.

Thank you!
Flavien

### Expected Behavior

The always_ff block should not set the value of the output signal to 1

### Actual Behavior

The always_ff block sets the value of the output signal to 1

**Comment by whitequark** [comment]
This is not directly relevant to the issue, but I think that if you want equivalence with Verilator, you should use one `step` for one `eval`.

**Comment by flaviens** [comment]
Thank you Whitequark for your response! Duplicating the the evals in Verilator does not change the outcome, as you may already have experimented.

**Comment by whitequark** [comment]
Yes. But it is something that is important for soundness of your testing technique.

In general, having two `step()` operations in a row in CXXRTL should be the same as having one. It's not universally a bug in CXXRTL but in many cases it is. Does it behave the same with just one?

**Comment by whitequark** [comment]
OK, that sounds like a CXXRTL bug.

**Comment by whitequark** [comment]
I've investigated this closer and there is this message from `proc_dff`:

```
3.3.9. Executing PROC_DFF pass (convert process syncs to FFs).
Creating register for signal `\wrapper_cxxrtl.\top_i.outreg' using process `\wrapper_cxxrtl.$flatten\top_i.$proc$top.sv:11$1'.
Warning: Async reset value `\in_0 [4]' is not constant!
  created $aldff cell `$procdff$3' with positive edge clock and negative level non-const reset.
```

I think this code isn't synthesizable Verilog:

```verilog
  always_ff @(posedge clkin_data[0], negedge clkin_data[32])
    if (!clkin_data[32]) outreg <= in_data[4];
    else outreg <= 1'b0;
```

Namely, this branch specifies a reset for an FF:

```
    if (!clkin_data[32]) outreg <= in_data[4];
```

However the reset value is not constant.

<hr>

I will request you not file any bugs against CXXRTL which have warnings emitted by the `proc_dff` pass.

**Comment by whitequark** [comment]
Here are the examples from IEEE 1364.1-2002:

![Screenshot_20240122_173519](https://github.com/YosysHQ/yosys/assets/54771/c71c5c52-e1ab-4245-bccd-3dd188ddb9ec)

**Comment by flaviens** [comment]
Thank you for your response and examples, this is very informative.
I will pay special attention to the proc_dff warnings if ever I get any other problems with CXXRTL.
Thank you for your time and for making CXXRTL available :+1:
